cmd/crank/beta/diff/client/kubernetes/resource_client.go

```diff
@@ -0,0 +1,171 @@
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	un "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+
+	"github.com/crossplane/crossplane/cmd/crank/beta/diff/client/core"
+)
+
+// ResourceClient handles basic CRUD operations for Kubernetes resources.
+type ResourceClient interface {
+	// GetResource retrieves a resource by its GVK, namespace, and name
+	GetResource(ctx context.Context, gvk schema.GroupVersionKind, namespace, name string) (*un.Unstructured, error)
+
+	// ListResources lists resources matching the given GVK and namespace
+	ListResources(ctx context.Context, gvk schema.GroupVersionKind, namespace string) ([]*un.Unstructured, error)
+
+	// GetResourcesByLabel returns resources matching labels in the given namespace
+	GetResourcesByLabel(ctx context.Context, namespace string, gvk schema.GroupVersionKind, sel metav1.LabelSelector) ([]*un.Unstructured, error)
+
+	// GetAllResourcesByLabels gets resources by labels across multiple GVKs
+	GetAllResourcesByLabels(ctx context.Context, gvks []schema.GroupVersionKind, selectors []metav1.LabelSelector) ([]*un.Unstructured, error)
+}
+
+// DefaultResourceClient implements the ResourceClient interface.
+type DefaultResourceClient struct {
+	dynamicClient   dynamic.Interface
+	discoveryClient discovery.DiscoveryInterface
+	converter       TypeConverter
+	logger          logging.Logger
+}
+
+// NewResourceClient creates a new DefaultResourceClient instance.
+func NewResourceClient(clients *core.Clients, converter TypeConverter, logger logging.Logger) ResourceClient {
+	return &DefaultResourceClient{
+		dynamicClient:   clients.Dynamic,
+		discoveryClient: clients.Discovery,
+		converter:       converter,
+		logger:          logger,
+	}
+}
+
+// GetResource retrieves a resource from the cluster based on its GVK, namespace, and name.
+func (c *DefaultResourceClient) GetResource(ctx context.Context, gvk schema.GroupVersionKind, namespace, name string) (*un.Unstructured, error) {
```

Comment:
Take a look at controller-runtime's client package. It's intended mostly for use in controllers but I've had good luck using it in CLIs. Similar to this package it's an abstraction on the underlying client-go clients. It supports cache backed clients already, so you might be able to avoid rolling your own.

https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.20.4/pkg/client